src/storage/store/mempool_poller.rs

```diff
@@ -0,0 +1,149 @@
+use crate::mempool::mempool::MempoolMessagesRequest;
+use crate::proto::{self, FarcasterNetwork, Transaction};
+use crate::storage::store::account::OnchainEventStorageError;
+use crate::utils::statsd_wrapper::StatsdClientWrapper;
+use itertools::Itertools;
+use std::string::ToString;
+use std::time::Duration;
+use thiserror::Error;
+use tokio::sync::{mpsc, oneshot};
+use tokio::time::timeout;
+use tracing::{error, info};
+
+#[derive(Error, Debug)]
+pub enum MempoolPollerError {
+    #[error(transparent)]
+    MessageReceiveError(#[from] oneshot::error::RecvError),
+
+    #[error(transparent)]
+    OnchainEventError(#[from] OnchainEventStorageError),
+}
+
+pub struct MempoolPoller {
+    pub messages_request_tx: Option<mpsc::Sender<MempoolMessagesRequest>>,
+    pub max_messages_per_block: u32,
+    pub network: FarcasterNetwork,
+    pub shard_id: u32,
+    pub statsd_client: StatsdClientWrapper,
+}
+
+#[derive(Clone, Debug)]
+pub enum MempoolMessage {
+    UserMessage(proto::Message),
+    ValidatorMessage(proto::ValidatorMessage),
+}
+
+impl MempoolMessage {
+    pub fn fid(&self) -> u64 {
+        match self {
+            MempoolMessage::UserMessage(msg) => msg.fid(),
+            MempoolMessage::ValidatorMessage(msg) => msg.fid(),
+        }
+    }
+
+    pub fn to_proto(&self) -> proto::MempoolMessage {
+        let msg = match self {
+            MempoolMessage::UserMessage(msg) => {
+                proto::mempool_message::MempoolMessage::UserMessage(msg.clone())
+            }
+            _ => todo!(),
+        };
+        proto::MempoolMessage {
+            mempool_message: Some(msg),
+        }
+    }
+}
+
+impl MempoolPoller {
+    pub fn for_read_node(&self) -> bool {
+        self.messages_request_tx.is_none()
```

Comment:
This has become a very roundabout way for the engine to determine if it's a read node. Might be fine for now, but wonder if we should just pass in a boolean to the engine instead.